test/Test/Assert.hs

```diff
@@ -0,0 +1,59 @@
+{- | We use @hspec@ testing framework and it doesn't support monad transformers.
+At this moment there's no testing framework that supports monad transformers. So
+we need to pass @AppEnv@ manually to every function.
+All functions take `AppEnv` as last argument. Like this one:
+@
+satisfies :: App a -> (a -> Bool) -> AppEnv -> Expectation
+@
+Because of that, there're multiple ways to write tests:
+@
+1. satisfies action isJust env
+2. (action `satisfies` isJust) env
+3. action `satisfies` isJust $ env
+4. env & action `satisfies` isJust
+@
+We can go even further and introduce fancy operators, so this code can be
+written in more concise way. But this is TBD.
+@
+env & action @? isJust
+@
+-}
+
+module Test.Assert
+       ( failsWith
+       , satisfies
+       , succeeds
+       , equals
+       ) where
+
+import Test.Hspec (Expectation, expectationFailure, shouldBe, shouldSatisfy)
+
+import IW.App (App, AppEnv, AppError (..), AppErrorType, runAppAsIO)
+
+
+-- | Checks that given action runs successfully.
+succeeds :: (Show a) => App a -> AppEnv -> Expectation
+succeeds = (`satisfies` const True)
+
+-- | Checks whether return result of the action satisfies given predicate.
+satisfies :: (Show a) => App a -> (a -> Bool) -> AppEnv -> Expectation
+satisfies app p env = runAppAsIO env app >>= \case
+    Left e  -> expectationFailure $ "Expected 'Success' but got: " <> show e
+    Right a -> a `shouldSatisfy` p
+
+-- | Checks whether action fails and returns given error.
+failsWith :: (Show a) => App a -> AppErrorType -> AppEnv -> Expectation
+failsWith app err env = runAppAsIO env app >>= \case
+    Left e@AppError{..} ->
+        if appErrorType == err
+            then appErrorType `shouldBe` err
+            else expectationFailure $
```

Comment:
This actually can be written simpler. I wrote it this way I don't remember the reason... @vrom911 noticed that here you can just write the following after `->`:

```haskell
appErrorType `shouldBe` err
```

If you have time to fix this in `three-layer` as well, it would be lovely :slightly_smiling_face:

Reply:
Got it!